src/lib/Names.js

```diff
@@ -0,0 +1,211 @@
+      }
+      // else
+      return false;
+    }
+
+    for (let i = 1; i < len; ++i) {
+      if (!isValidCharInName(name[i])) {
+        if (throwError) {
+          throw new Error(`Character [${c}] at element [${i}] is not valid in Graph Resource Name [${name}].  Valid characters are a-z, A-Z, 0-9, / and _.`);
+        }
+        // else
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  clean(name) {
+    name = name.replace(CLEAN_REGEX, '/');
+
+    if (name.endsWith('/')) {
+      return name.substr(0, -1);
+    }
+    // else
+    return name;
+  }
+
+  append(left, right) {
+    return this.clean(left + '/' + right);
+  }
+
+  remap(name) {
+    return this.resolve(name, true);
+  }
+
+  /**
+   * @param [namespace] {string} namespace to resolve name to. If not provided, node's namespace will be used
+   * @param name {string} name to resolve
+   * @param [remap] {bool} flag indicating if we should also attempt to remap the name
+   */
+  resolve(...args) {
+    let [namespace, name, remap] = this._parseResolveArgs(args);
+
+    this.validate(name, true);
+
+    if (name.length === 0) {
+      if (namespace.length === 0) {
+        return '/';
+      }
+      else if (namespace[0] === '/') {
+        return namespace
+      }
+      // else
+      return '/' + namespace;
+    }
+
+    if (name.startsWith('~')) {
+      name = name.replace('~', this._namespace + '/');
+    }
+
+    if (!name.startsWith('/')) {
+      name = namespace + '/' + name;
+    }
+
+    name = this.clean(name);
+
+    if (remap) {
+      name = this._remap(name);
+    }
+
+    return name;
+  }
+
+  parentNamespace(name) {
+    this.validate(name, true);
+
+    if (name.length === 0) {
+      return '';
+    }
+    else if (name === '/') {
+      return '/';
+    }
+
+    let p = name.lastIndexOf('/');
+    if (p === name.length - 1) {
+      p = name.lastIndexOf('/', p - 1);
+    }
+
+    if (p < 0) {
+      return '';
+    }
+    else if (p === 0) {
+      return '/';
+    }
+    // else
+    return name.substring(0, p);
+  }
+
+  _remap(name) {
+    return this._remappings[name] || name;
+  }
+
+  _parseResolveArgs(args) {
+    let name, namespace = this._namespace, remap = true;
+    switch (args.length) {
+      case 0:
+        name = '';
+        break;
+      case 1:
+        name = args[0];
+        break;
+      case 2:
+        if (typeof args[1] === 'string') {
+          [namespace, name] = args;
+        }
+        else {
+          [name, remap] = args;
+        }
+        break;
+      default:
+        [namespace, name, remap] = args;
+        break;
+    }
+
+    return [namespace, name, remap];
+  }
+}
+
+module.exports = new Names();
+
+//------------------------------------------------------------------
+// Local Helper functions
+//------------------------------------------------------------------
+
+
+
+function isAlpha(char, code = null) {
+  code = code || char.charCodeAt(0);
+  if (code >= 65 && code <= 90) {
```

Comment:
JS char comparison already uses charCode internally so technically we can do it like this:
```
char >= 'A' && char <= 'Z'
```

I did a jsperf and it's faster than do charCodeAt:

![image](https://user-images.githubusercontent.com/6723574/42255938-30beda9c-7f03-11e8-81f6-f08d22ef18e2.png)

DYI at: https://jsperf.com/char-code-comparison/1